src/filesys.cpp

```diff
@@ -755,49 +757,64 @@ const char *GetFilenameFromPath(const char *path)
 
 bool safeWriteToFile(const std::string &path, const std::string &content)
 {
-	std::string tmp_file = path + ".~mt";
+	// Write to tmp file
+	std::random_device rd;
+	std::mt19937 gen(rd());
+	std::uniform_int_distribution<unsigned int> distrib(0);
 
-	// Write to a tmp file
-	std::ofstream os(tmp_file.c_str(), std::ios::binary);
-	if (!os.good())
-		return false;
-	os << content;
-	os.flush();
-	os.close();
-	if (os.fail()) {
-		// Remove the temporary file because writing it failed and it's useless.
-		remove(tmp_file.c_str());
+	int random_int = distrib(gen);
+
+	std::stringstream ss;
+	ss << std::hex << random_int;
+
+	std::string tmp_path = TempPath() + DIR_DELIM + "minetest_" + ss.str();
+	std::ofstream tmp_os(tmp_path.c_str(), std::ios::binary);
+	if (!tmp_os.good()) {
+		warningstream << "Failed to create temp file: " << tmp_path << std::endl;
 		return false;
 	}
 
-	bool rename_success = false;
+	{ // scope for the unique_ptr
+		// Attempt to remove the temporary file on exit, but don't worry about checking
+		// success since the OS should clean it up anyway.
+		auto deleter = [&tmp_path](std::ofstream *os) { std::remove(tmp_path.c_str()); };
+		std::unique_ptr<std::ofstream, decltype(deleter)> tmp_os_ptr(&tmp_os, deleter);
+		tmp_os << content;
+		tmp_os.flush();
+		tmp_os.close();
+		if (tmp_os.fail()) {
+			warningstream << "Failed to write to temp file: " << tmp_path << std::endl;
+			return false;
+		}
 
-	// Move the finished temporary file over the real file
+		bool write_succeeded = false;
 #ifdef _WIN32
-	// When creating the file, it can cause Windows Search indexer, virus scanners and other apps
-	// to query the file. This can make the move file call below fail.
-	// We retry up to 5 times, with a 1ms sleep between, before we consider the whole operation failed
-	int number_attempts = 0;
-	while (number_attempts < 5) {
-		rename_success = MoveFileEx(tmp_file.c_str(), path.c_str(),
-				MOVEFILE_REPLACE_EXISTING | MOVEFILE_WRITE_THROUGH);
-		if (rename_success)
-			break;
-		sleep_ms(1);
-		++number_attempts;
-	}
+		// On Windows we've observed that the newly created temp file is often
+		// being read by other processes and Windows does not allow moving
+		// of a file while it is being read, so we retry and then fall back to copying.
+		// NOTE: once Windows 10 is the lowest supported verison, we should consider
+		// using FILE_DISPOSITION_POSIX_SEMANTICS to avoid this issue.
```

Comment:
I think I found some other project that used it with rename, but I don't think it's worth the trouble to find it now, since this code has to support Windows 8 still. I can remove the comment if you want.